src/linear_algebra/nonsingular_inverse.lean

```diff
@@ -354,33 +354,77 @@ Defines the matrix `nonsing_inv A` and proves it is the inverse matrix
 of a square matrix `A` as long as `det A` has a multiplicative inverse.
 -/
 
-variables [comm_ring α] [has_inv α]
+variables [comm_ring α]
+variables (A : matrix n n α)
 
-/-- The inverse of a nonsingular matrix.
+open_locale classical
 
-  This is not the most general possible definition, so we don't instantiate `has_inv` (yet).
--/
-def nonsing_inv (A : matrix n n α) : matrix n n α := (A.det)⁻¹ • adjugate A
+lemma is_unit_det_transpose (h : is_unit A.det) : is_unit Aᵀ.det :=
+by { rw det_transpose, exact h, }
+
+/-- The inverse of a square matrix, when it is invertible (and zero otherwise).-/
+noncomputable def nonsing_inv : matrix n n α :=
+if h : is_unit A.det then (↑h.unit⁻¹ : α) • A.adjugate else 0
```

Comment:
If you change `else 0` to `else 1`, then the inverse is always invertible. I haven't thought of other pros/cons of `0` vs `1`.

Reply:
For one-dimensional matrices, I want the inverse to coincide with the field inverse of the unique coefficient of the matrix. So, please, keep 0 :-)

Reply:
Good point.